Improved cookieStoreId documentation (#29337)
mdn/content: files/en-us/mozilla/add-ons/webextensions/work_with_contextual_identities/index.md

## APIs for working with contextual identities

> To provide contextual identity features in an extension, you use {{WebExtAPIRef("contextualIdentities")}} to add, query, update, and delete contextual identities. When you create a contextual identity you obtain its `cookieStoreId`. You can then use the cookie store ID to work with entities related to the conceptual identity using these APIs:

Author: @Rob--W "conceptual identity" was unintended

- {{WebExtAPIRef("contentscripts.register")}} which enables you register a content script in the tabs that belong to one or more cookie store IDs.
- {{WebExtAPIRef("downloads")}} where you can associate a download with a cookie store.
- {{WebExtAPIRef("proxy")}} where the details passed into the {{WebExtAPIRef("proxy.onRequest")}} listener identify the cookie store associate the request.
- {{WebExtAPIRef("tabs")}} where you can {{WebExtAPIRef("tabs.create","create")}} a tab to it uses a contextual identity's cookies store, {{WebExtAPIRef("tabs.tab","get")}} the cookie store ID for a tab, and {{WebExtAPIRef("tabs.query","query")}} tabs based on their associated cookie store.
- {{WebExtAPIRef("userscripts.register")}} which enables you to register a content script in the tabs that belong to one or more cookie store IDs.
- {{WebExtAPIRef("webrequest")}} where all the events return the cookie store ID if request is associated with a contextual identities tab.
- {{WebExtAPIRef("windows.create")}} where you can specify the cookie store for the tabs added to a window when its created.

## Permissions

> To use the {{WebExtAPIRef("contextualIdentities")}} API you need to include the "contextualIdentities" [permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) in your [manifest.json](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json) file. You don't need the "tabs" permission to use {{WebExtAPIRef("tabs.create")}}.

Reviewer: ```suggestion
To use the {{WebExtAPIRef("contextualIdentities")}} API you need to include the "contextualIdentities" [permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) in your [manifest.json](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json) file.
```

The mention of the "tabs" permission is out of place here, especially with the new structure of this section.

> If an API enables cookies to be modified then you need the "cookies" permission. For example, in {{WebExtAPIRef("tabs.query")}}, use of`cookieStoreId` does not require the "cookies" API as reading the property doesn't affect the cookies in the containers. However, using {{WebExtAPIRef("tabs.create")}} requires the permission, because it enables the creation of cookies in a container.

Reviewer: ```suggestion
If an API enables cookies to be modified then you need the "cookies" permission. For example, in {{WebExtAPIRef("tabs.query")}}, use of`cookieStoreId` does not require the "cookies" API as reading the property doesn't affect the cookies in the containers. However, using {{WebExtAPIRef("tabs.create")}} requires the permission, because the opened tab can read and modify cookies in a container.
```

## APIs for working with contextual identities

Depending on the nature of your extension, you may want to manage contextual identities, associate objects your extension manipulates with contextual identities, or both.

### Managing contextual identities

To manage contextual identities, you use the {{WebExtAPIRef("contextualIdentities")}} API. This API enables you to add, query, update, and delete contextual identities. When you create a contextual identity, it is given a unique `cookieStoreId`. You use this ID to work with entities related to the contextual identity.

### Using `cookieStoreId`

Several extension APIs include the `cookieStoreId` in objects to enable extensions to associate these objects with specific contextual identities.

- {{WebExtAPIRef("browsingData.removeCookies()")}} and {{WebExtAPIRef("browsingData.removeLocalStorage()")}} where you use {{WebExtAPIRef("browsingData.removalOptions")}} to set the cookie store items are removed from.

> - {{WebExtAPIRef("contentscripts.register")}} enables you to register a content script in the tabs belonging to one or more cookie store IDs.

Reviewer: ```suggestion
- {{WebExtAPIRef("contentscripts.register")}} enables you to register a content script restricted to documents associated with one or more specific cookieStoreIds.
```

- {{WebExtAPIRef("downloads")}} where you can associate a download with a cookie store.
- {{WebExtAPIRef("proxy")}} where the details passed into the {{WebExtAPIRef("proxy.onRequest")}} listener identify the cookie store associated with the request.

> - {{WebExtAPIRef("tabs")}} where you can {{WebExtAPIRef("tabs.create","create")}} a tab to use a contextual identity's cookies store, {{WebExtAPIRef("tabs.tab","get")}} the cookie store ID for a tab, and {{WebExtAPIRef("tabs.query","query")}} tabs based on their associated cookie store.

Reviewer: ```suggestion
- {{WebExtAPIRef("tabs")}} where you can {{WebExtAPIRef("tabs.create","create")}} a tab in a specific container tab, {{WebExtAPIRef("tabs.tab","get")}} the cookieStoreId for a tab, and {{WebExtAPIRef("tabs.query","query")}} tabs based on their associated cookie store.
```

cookieStoreId consists of container tabs, private browsing context and the default (non-private) browsing mode. `tabs.create` cannot be used to open a private browsing tab in a non-private window, or a non-private tab in a private window. Therefore I'm specifically pointing to "container tabs" in the tabs.create method.

> - {{WebExtAPIRef("userscripts.register")}} enables you to register a content script in the tabs belonging to one or more cookie store IDs.

Reviewer: ```suggestion
- {{WebExtAPIRef("userscripts.register")}} enables you to register a content script restricted to documents associated with one or more specific cookieStoreIds.
```

> - {{WebExtAPIRef("webrequest")}} where all the events return the cookie store ID if the request is associated with a contextual identities tab.

Reviewer: ```suggestion
- {{WebExtAPIRef("webrequest")}} where all the events return the cookieStoreId of the request.
```

Also for non-container tabs (and even non-tab requests such as service workers).
